connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java

```diff
@@ -0,0 +1,121 @@
+        public static final String CONNECTOR_SCOPE = "connector.scope";
+    }
+
+    /**
+     * The {@link Parameters#CONNECTOR_SCOPE scope} values used by Connect.
+     */
+    public enum Scope {
+        /**
+         * The {@link Parameters#CONNECTOR_SCOPE scope} value for the worker as it starts a connector.
+         */
+        WORKER("worker"),
+
+        /**
+         * The {@link Parameters#CONNECTOR_SCOPE scope} value for {@link org.apache.kafka.connect.connector.Task}
+         * implementations.
+         */
+        TASK("task"),
+
+        /**
+         * The {@link Parameters#CONNECTOR_SCOPE scope} value for committing offsets.
+         */
+        OFFSETS("offsets");
+
+        private final String value;
+        Scope(String value) {
+            this.value = value;
+        }
+        public String value() {
+            return this.value;
+        }
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the {@link Parameters#CONNECTOR_NAME connector name} to the
+     * supplied name and the {@link Parameters#CONNECTOR_SCOPE scope} to {@link Scope#WORKER}.
+     *
+     * @param connectorName the connector name; may not be null
+     */
+    public static void forConnector(String connectorName) {
+        MDC.put(Parameters.CONNECTOR_NAME, connectorName);
+        MDC.remove(Parameters.CONNECTOR_TASK);
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.WORKER.value());
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the {@link Parameters#CONNECTOR_NAME connector name} and
+     * {@link Parameters#CONNECTOR_TASK task number} to the same values in the supplied {@link ConnectorTaskId}.
+     *
+     * @param id the connector task ID; may not be null
+     */
+    public static void forTask(ConnectorTaskId id) {
+        MDC.put(Parameters.CONNECTOR_NAME, id.connector());
+        MDC.put(Parameters.CONNECTOR_TASK, Integer.toString(id.task()));
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.TASK.value());
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the  {@link Parameters#CONNECTOR_NAME connector name},
+     * {@link Parameters#CONNECTOR_TASK task number} and to the same values in the supplied {@link ConnectorTaskId}, and to set
+     * the scope to {@link Scope#OFFSETS}.
+     *
+     * @param id the connector task ID; may not be null
+     */
+    public static void forOffsets(ConnectorTaskId id) {
+        MDC.put(Parameters.CONNECTOR_NAME, id.connector());
+        MDC.put(Parameters.CONNECTOR_TASK, Integer.toString(id.task()));
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.OFFSETS.value());
+    }
+
+    /**
+     * Remove all of the Connect-specific {@link Parameters parameters} from the current {@link MDC} logging context.
+     */
+    public static void clear() {
+        MDC.remove(Parameters.CONNECTOR_NAME);
+        MDC.remove(Parameters.CONNECTOR_TASK);
+        MDC.remove(Parameters.CONNECTOR_SCOPE);
```

Comment:
just wondering why we need this method? AFAICT, we don't reuse threads in Connect. If a thread is dedicated for a particular task (as a task or offset committer), it is dedicated for that task throughout its lifetime. if we could get away without needing this method, it would simplify the changes. WDYT?

Reply:
You're right that it's not often necessary. Where we do not switch threads, the extra overhead of resetting the MDC parameters is negligible. But the offset commit thread does switch logging contexts, and IMO it is simply a good practice to encourage restoring resources. 

Per @ewencp's suggestion, I went back to AutoCloseable so that try-with-resources works and reduces the risk of coding errors.

Reply:
The heavy lifting in `SourceTaskOffsetCommitter` is delegated back to the WorkerSourceTask though, which doesn't help us with MDC but does bring up if we want to use LogContext instead of MDC for AK? Other components in AK (Kafka, Streams and Clients) seem to use [LogContext](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/LogContext.java#L37-L39) that adds a prefix on every SLF4J logger. This is already being used by some bit of Connect (for example., in WorkerCoordinator). This would avoid having to micromanage thread context, and also keeps the platform uniform.

Reply:
Yes, I had the same thought as @wicknicks. We preferred `LogContext` over MDC because our needs were basic and the API was simpler.

Reply:
@wicknicks, @hachikuji: indeed `LogContext` is simpler, but:
1. its prefix is fixed; and
2. it only affects log messages written by `Logger` instances instantiated with the `LogContext`.

So for a Kafka client, this works great because the client has a single ID and the AK code that creates the `Logger` can be changed to use `LogContext`.

However, with Connect the connector implementations will not change how they create their `Logger` instances. And, we do have some code where a single thread does switch contexts from one connector/task to another. Given these, I don't see how `LogContext` will work. 

This PR uses MDC, and this will add the additional context (via MDC parameters) to _all_ log messages written by _all_ components -- including connector implementations. MDC works this way because it is thread-based, not based upon the `Logger` instance that is being used, so switching contexts within a single thread is easy to do.

Reply:
alright cool -- yeah LogContext won't work if we want to tag message from plugins.